libs/SalesforceSDKCore/SalesforceSDKCore/Classes/Login/SFLoginViewController+QrCodeLogin.swift

```diff
@@ -0,0 +1,169 @@
+//  permission of salesforce.com, inc.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
+//  IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+//  FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+//  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+//  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
+//  WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import Foundation
+
+public extension SalesforceLoginViewController {
+
+    // MARK: - QR Code Login Via UI Bridge API Public Implementation
+
+    /**
+     * Automatically log in with a UI Bridge API login QR code.
+     * - Parameters
+     *     - loginQrCodeContent:  The login QR code content.  This should be either a URL or URL
+     *     query containing the UI Bridge API JSON parameter.  The UI Bridge API JSON parameter
+     *     should contain URL-encoded JSON with two values:
+     *       - frontdoor_bridge_url
+     *       - pkce_code_verifier
+     * If pkce_code_verifier is not specified then the user agent flow is used
+     * - Returns: Boolean true if a log in attempt is possible using the provided QR code content, false
+     * otherwise
+     */
+    func loginFromQrCode(
+        loginQrCodeContent: String?
+    ) -> Bool {
+        if let uiBridgeApiParameters = uiBridgeApiParametersFromLoginQrCodeContent(
+            loginQrCodeContent
+        ) {
+            loginWithFrontdoorBridgeUrl(
+                uiBridgeApiParameters.frontdoorBridgeUrl,
+                pkceCodeVerifier: uiBridgeApiParameters.pkceCodeVerifier
+            )
+            return true
+        } else {
+            return false
+        }
+    }
+
+    /**
+     * Automatically log in with a UI Bridge API front door bridge URL and PKCE code verifier.
+     * - Parameters
+     *   - frontdoorBridgeUrl: The UI Bridge API front door bridge URL
+     *   - pkceCodeVerifier: The PKCE code verifier
+     */
+    func loginWithFrontdoorBridgeUrl(
+        _ frontdoorBridgeUrlString: String,
+        pkceCodeVerifier: String?
+    ) {
+        guard let frontdoorBridgeUrl = URL(string: frontdoorBridgeUrlString) else { return }
+        
+        // Stop current authentication attempt, if applicable, before starting the new one.
+        UserAccountManager.shared.stopCurrentAuthentication { result in
+            
+            DispatchQueue.main.async {
+                // Login using front door bridge URL and PKCE code verifier provided by the QR code.
+                AuthHelper.loginIfRequired(nil,
```

Comment:
Note the new `AuthHelp` convenience here with parameters supporting QR code login.